File: src/main/java/com/anywherecommerce/android/sdk/sampleapp/MainActivity.java
package com.anywherecommerce.android.sdk.sampleapp;

import android.app.Activity;
import android.bluetooth.BluetoothDevice;
import android.graphics.Color;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.anywherecommerce.android.sdk.AppBackgroundingManager;
import com.anywherecommerce.android.sdk.AuthenticationListener;
import com.anywherecommerce.android.sdk.Endpoint;
import com.anywherecommerce.android.sdk.Logger;
import com.anywherecommerce.android.sdk.MeaningfulError;
import com.anywherecommerce.android.sdk.MeaningfulErrorListener;
import com.anywherecommerce.android.sdk.MeaningfulMessage;
import com.anywherecommerce.android.sdk.SDKManager;
import com.anywherecommerce.android.sdk.Terminal;
import com.anywherecommerce.android.sdk.component.SignatureView;
import com.anywherecommerce.android.sdk.devices.BluetoothCardReaderConnectionListener;
import com.anywherecommerce.android.sdk.devices.CardReaderConnectionListener;
import com.anywherecommerce.android.sdk.devices.CardReaderController;
import com.anywherecommerce.android.sdk.devices.TransactionWorkflow;
import com.anywherecommerce.android.sdk.devices.USBCardReaderConnectionListener;
import com.anywherecommerce.android.sdk.devices.bbpos.BBPOSDevice;
import com.anywherecommerce.android.sdk.devices.bbpos.BBPOSDeviceCardReaderController;
import com.anywherecommerce.android.sdk.devices.CardReader;
import com.anywherecommerce.android.sdk.devices.MultipleBluetoothDevicesFoundListener;
import com.anywherecommerce.android.sdk.RequestListener;
import com.anywherecommerce.android.sdk.GenericEventListener;
import com.anywherecommerce.android.sdk.GenericEventListenerWithParam;
import com.anywherecommerce.android.sdk.devices.bbpos.WalkerC2X;
import com.anywherecommerce.android.sdk.endpoints.AnyPayTransaction;
import com.anywherecommerce.android.sdk.endpoints.prioritypayments.PriorityPaymentsEndpoint;
import com.anywherecommerce.android.sdk.endpoints.propay.PropayEndpoint;

import com.anywherecommerce.android.sdk.endpoints.worldnet.WorldnetEndpoint;
import com.anywherecommerce.android.sdk.models.CustomerDetails;
import com.anywherecommerce.android.sdk.models.TipLineItem;
import com.anywherecommerce.android.sdk.models.TransactionType;

import com.anywherecommerce.android.sdk.transactions.CardTransaction;
import com.anywherecommerce.android.sdk.transactions.Transaction;
import com.anywherecommerce.android.sdk.transactions.listener.CardTransactionListener;
import com.anywherecommerce.android.sdk.transactions.listener.TransactionListener;
import com.anywherecommerce.android.sdk.util.Amount;

import java.util.List;


/**
 * Created by Admin on 10/4/2017.
 */

public class MainActivity extends Activity {

    TextView txtPanel;
    EditText txtReferenceId;
    Button btUSBConnect, btnConnectAudio, btnDisconnectAudio, btnIsDeviceConnected, btnStartEMV, btnConnectBT, btnDisconnectBT, btnGetTransactions, btnTerminalLogin, btnKeyedsale;
    Button btnCaptureTransaction, btnRefRefund;
    DialogManager dialogs = new DialogManager();
    Transaction refTransaction;
    Endpoint endpoint;
    String propaySessionID;
    Button btnSubmitSignature;
    LinearLayout signatureViewLayout;
    SignatureView signatureView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.testharness_device);

        if (!PermissionsController.verifyAppPermissions(this)) {
            PermissionsController.requestAppPermissions(this, PermissionsController.permissions, 1001);
        }

        Logger.i("SDK Version - " + SDKManager.getSdkVersion());

        Terminal.initialize();
        endpoint = Terminal.getInstance().getEndpoint();

        AppBackgroundingManager.get().registerListener(new AppBackgroundingManager.AppBackroundedListener() {
            @Override
            public void onBecameForeground() {
                Logger.trace("Caught app in foreground.");
            }

            @Override
            public void onBecameBackground() {
                Logger.trace("Caught app in background.");
            }
        });

        txtPanel = (TextView) findViewById(R.id.txtTextHarnessPanel);
        txtPanel.setMovementMethod(new ScrollingMovementMethod());

        btnConnectAudio = (Button) findViewById(R.id.audioConnect);
        btnDisconnectAudio = (Button) findViewById(R.id.btnDisconnectAudio);
        btnConnectBT = (Button) findViewById(R.id.btConnect);
        btnDisconnectBT = (Button) findViewById(R.id.btnDisconnectBT);
        btnIsDeviceConnected = (Button) findViewById(R.id.btnIsDeviceConnected);
        btnGetTransactions = (Button) findViewById(R.id.btnGetTransactions);
        btnStartEMV = (Button) findViewById(R.id.btnStartEmvSale);
        btnTerminalLogin = (Button) findViewById(R.id.terminalLogin);
        btnKeyedsale = (Button) findViewById(R.id.keyedsale);
        btnCaptureTransaction = (Button) findViewById(R.id.unrefRefund);
        btnRefRefund = (Button) findViewById(R.id.refRefund);
        txtReferenceId = (EditText) findViewById(R.id.txtReferenceId);
        btUSBConnect = (Button) findViewById(R.id.btUSBConnect);
        signatureViewLayout = (LinearLayout) findViewById(R.id.signatureViewLayout);
        signatureView = (SignatureView) findViewById(R.id.signatureView);
        btnSubmitSignature = (Button) findViewById(R.id.btnSubmitSignature);

        signatureView.setStrokeColor(Color.MAGENTA);
        signatureView.setStrokeWidth(10f);


        final CardReaderController cardReaderController = CardReaderController.getControllerFor(BBPOSDevice.class);

        cardReaderController.subscribeOnCardReaderConnected(new GenericEventListenerWithParam<CardReader>() {
            @Override
            public void onEvent(CardReader deviceInfo) {
                if (deviceInfo == null)
                    addText("\r\nUnknown device connected");
                else
                    addText("\nDevice connected " + deviceInfo.getModelDisplayName());
            }
        });

        cardReaderController.subscribeOnCardReaderDisconnected(new GenericEventListener() {
            @Override
            public void onEvent() {
                addText("\nDevice disconnected");
            }
        });

        cardReaderController.subscribeOnCardReaderConnectFailed(new MeaningfulErrorListener() {
            @Override
            public void onError(MeaningfulError error) {
                addText("\nDevice connect failed: " + error.toString());
            }
        });

        cardReaderController.subscribeOnCardReaderError(new MeaningfulErrorListener() {
            @Override
            public void onError(MeaningfulError error) {
                addText("\nDevice error: " + error.toString());
            }
        });

        btUSBConnect.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("\nConnecting to USB Reader\r\n");
                cardReaderController.connectOther(CardReader.ConnectionMethod.USB);
            }
        });

        btnConnectAudio.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("\nConnecting to audio jack (with polling)\r\n");
                cardReaderController.connectAudioJack();
            }
        });

        btnDisconnectAudio.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("\r\nNot implemented");
                //addText("Disconnecting audio jack\r\n");
                //BBPOSDeviceCardReaderController.getInstance().disconnectAudioJack();
            }
        });

        btnConnectBT.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("\nConnecting to BT\r\n");
                cardReaderController.connectBluetooth(new MultipleBluetoothDevicesFoundListener() {
                    @Override
                    public void onMultipleBluetoothDevicesFound(List<BluetoothDevice> matchingDevices) {
                        addText("Many BT devices");
                    }
                });
            }
        });

        btnDisconnectBT.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("Disconnecting bluetooth\r\n");
                cardReaderController.disconnectReader();
            }
        });


        btnIsDeviceConnected.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addText("\nClicky");
                BBPOSDeviceCardReaderController.getController().setDebugLogEnabled(true);
                if (latch) {
                    addText("\r\nStopping audio");
                    cardReaderController.disconnectReader();
                } else {
                    addText("\r\nStarting audio");
                    cardReaderController.connectAudioJack();
                }

                latch = !latch;
            }
        });

        btnStartEMV.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                BBPOSDeviceCardReaderController.getController().setDebugLogEnabled(true);

                addText("\nStarting EMV transaction");
                if (!CardReaderController.isCardReaderConnected()) {
                    addText("\r\nNo card reader connected");
                    return;
                }

                dialogs.showProgressDialog(MainActivity.this, "Please Wait...");

                sendEMVTransaction();
            }
        });



        btnTerminalLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                authenticateTerminal();
            }
        });

        btnKeyedsale.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addText("\r\nExecuting Keyed Sale Transaction. Please Wait...");
                dialogs.showProgressDialog(MainActivity.this, "Please Wait...");

                sendKeyedTransaction();
            }

            ;
        });


        btnCaptureTransaction.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {


                addText("----> Starting New Refund Transaction <----");
                dialogs.showProgressDialog(MainActivity.this, "Please Wait...");

                final AnyPayTransaction transaction = new AnyPayTransaction();
                transaction.setEndpoint(endpoint);
                transaction.setTransactionType(TransactionType.REFUND);
                transaction.setTotalAmount(new Amount("20.25"));
                transaction.setCardExpiryMonth("10");
                transaction.setCardExpiryYear("20");
                transaction.setAddress("123 Main Street");
                transaction.setPostalCode("30004");
                transaction.setCVV2("999");
                transaction.setCardholderName("Jane Doe");
                transaction.setCardNumber("4012888888881881");
                transaction.execute(new TransactionListener() {
                    @Override
                    public void onTransactionCompleted() {
                        dialogs.hideProgressDialog();

                        if (transaction.isApproved())
                            MainActivity.this.addText("----> Transaction Refunded <----");
                        else {
                            MainActivity.this.addText(transaction.getResponseText());
                        }
                    }

                    @Override
                    public void onTransactionFailed(MeaningfulError meaningfulError) {
                        dialogs.hideProgressDialog();

                        MainActivity.this.addText("Refund Failed");
                    }
                });
            }
        });

        btnRefRefund.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (refTransaction == null) {
                    addText("----> This voids the last Auth transaction made. Please perform a Auth transaction first. <----");
                    return;
                }

                addText("----> Starting Referenced Refund Transaction <----");

                final AnyPayTransaction transaction = new AnyPayTransaction();
                transaction.setEndpoint(endpoint);
                transaction.setExternalId(refTransaction.getExternalId());
                transaction.setTotalAmount(new Amount("1"));
                transaction.setRefTransactionId(refTransaction.getExternalId());
                transaction.setTransactionType(TransactionType.REFUND);

                //t.enableLogging();
                transaction.execute(new TransactionListener() {

                    @Override
                    public void onTransactionCompleted() {
                        dialogs.hideProgressDialog();

                        if (transaction.isApproved())
                            MainActivity.this.addText("----> Transaction Refunded <----");
                        else {
                            MainActivity.this.addText(transaction.getResponseText());
                        }

                    }

                    @Override
                    public void onTransactionFailed(MeaningfulError reason) {
                        dialogs.hideProgressDialog();

                        MainActivity.this.addText("Refund Failed");
                    }
                });
            }
        });

        btnSubmitSignature.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dialogs.showProgressDialog(MainActivity.this, "Please Wait...");
                ((CardTransaction)refTransaction).setSignature(signatureView.getSignature());

                if (refTransaction.getCustomField("completed") != null) {

                    ((WorldnetEndpoint)endpoint).submitSignature(refTransaction, new RequestListener() {
                        @Override
                        public void onRequestComplete(Object o) {
                            addText("\r\n SIgnature Sent Successfully");
                        }

                        @Override
                        public void onRequestFailed(MeaningfulError meaningfulError) {
                            addText("\r\n SIgnature sent Failed");
                        }
                    });
                }
                else {
                    refTransaction.proceed();
                }

                addText("\r\nSending SIgnature");

            }
        });

        authenticateTerminal();
    }


    private void sendKeyedTransaction() {
        final AnyPayTransaction transaction = new AnyPayTransaction();
        transaction.setEndpoint(endpoint);
        transaction.setTransactionType(TransactionType.SALE);
        transaction.setCardExpiryMonth("10");
        transaction.setCardExpiryYear("20");
        transaction.setAddress("123 Main Street");
        transaction.setPostalCode("30004");
        transaction.setCVV2("999");
        transaction.setCardholderName("Jane Doe");
        transaction.setCardNumber("4012888888881881");
        transaction.setTotalAmount(new Amount("10.47"));
        transaction.setCurrency("USD");

        refTransaction = transaction;
        //t.enableLogging();
        transaction.execute(transactionListener);

    }

    private void sendEMVTransaction() {
        final AnyPayTransaction transaction = new AnyPayTransaction();
        transaction.setEndpoint(endpoint);
        transaction.useCardReader(CardReaderController.getConnectedReader());
        transaction.setTransactionType(TransactionType.SALE);
        transaction.setAddress("123 Main Street");
        transaction.setPostalCode("30004");
        transaction.setTotalAmount(new Amount("10.47"));
        transaction.setCurrency("USD");

        transaction.setOnSignatureRequiredListener(new GenericEventListener() {
            @Override
            public void onEvent() {
                addText("\r\n------>onSignatureRequired: sending null and proceeding");
                String signature = "base64-encoded image or point map";
                dialogs.hideProgressDialog();

                signatureViewLayout.setVisibility(View.VISIBLE);
            }
        });

        refTransaction = transaction;

        //t.enableLogging();
        transaction.execute(new CardTransactionListener() {
            @Override
            public void onCardReaderEvent(MeaningfulMessage event) {
                addText("\r\n------>onCardReaderEvent: " + event.message);
            }

            @Override
            public void onTransactionCompleted() {
                dialogs.hideProgressDialog();
                addText("\r\n------>onTransactionCompleted" + transaction.isApproved().toString());

                signatureViewLayout.setVisibility(View.GONE);
            }

            @Override
            public void onTransactionFailed(MeaningfulError reason) {
                dialogs.hideProgressDialog();
                addText("\r\n------>onTransactionFailed: " + reason.toString());

                signatureViewLayout.setVisibility(View.GONE);
            }
        });
    }

    static boolean latch = false;

    private void addText(String text) {
        txtPanel.append("\r\n" + text + "\r\n\n");
    }

    private void authenticateTerminal() {

        addText("Authenticating. Please Wait...");


        ((WorldnetEndpoint)endpoint).setWorldnetSecret("");
        ((WorldnetEndpoint)endpoint).setWorldnetTerminalID("");

        ((WorldnetEndpoint)endpoint).setUrl("https://testpayments.anywherecommerce.com/merchant");


        ((WorldnetEndpoint)endpoint).authenticate(new AuthenticationListener() {
            @Override
            public void onAuthenticationComplete() {
                addText("\r\n------> Terminal Authenticated");
            }

            @Override
            public void onAuthenticationFailed(MeaningfulError meaningfulError) {
                addText("\r\n------> Terminal Authentication Failed");
            }
        });

    }

    private TransactionListener transactionListener = new TransactionListener() {
        @Override
        public void onTransactionCompleted() {
            dialogs.hideProgressDialog();
            addText("\r\n------>onTransactionCompleted - " + refTransaction.isApproved().toString());

            if (refTransaction.getTransactionType() == TransactionType.SALE) {
                final TipLineItem tip = new TipLineItem();
                tip.amount = new Amount("23");

                refTransaction.setTip(tip);

                ((WorldnetEndpoint)endpoint).submitTipAdjustment(refTransaction, tip, new RequestListener() {
                    @Override
                    public void onRequestComplete(Object o) {
                        addText("\r\n------>Tip Adjustment success");
                    }

                    @Override
                    public void onRequestFailed(MeaningfulError meaningfulError) {
                        addText("\r\n------>Tip Adjustment failed");
                    }
                });

                /*((AnyPayTransaction)refTransaction).update(new TransactionListener() {
                    @Override
                    public void onTransactionCompleted() {
                        addText("\r\n------>update transaction complete - " + refTransaction.isApproved().toString());

                        CustomerDetails customerDetails = new CustomerDetails();
                        customerDetails.setEmailAddress("theiosdevguy@gmail.com");
                        refTransaction.setCustomerDetails(customerDetails);
                        ((AnyPayTransaction) refTransaction).update(new TransactionListener() {
                            @Override
                            public void onTransactionCompleted() {
                                addText("\r\n------>Receipt Sent - ");

                                refTransaction.addCustomField("completed", true);
                                signatureViewLayout.setVisibility(View.VISIBLE);

                            }

                            @Override
                            public void onTransactionFailed(MeaningfulError meaningfulError) {
                                addText("\r\n------>Receipt Sent Failed ");
                            }
                        });
                    }

                    @Override
                    public void onTransactionFailed(MeaningfulError meaningfulError) {
                        addText("\r\n------>update transaction failed - " + refTransaction.isApproved().toString());
                    }
                });*/
            }
        }

        @Override
        public void onTransactionFailed(MeaningfulError reason) {
            dialogs.hideProgressDialog();
            addText("\r\n------>onTransactionFailed: " + reason.toString());
        }
    };
}

File: src/main/java/com/anywherecommerce/android/sdk/sampleapp/MainActivity.kt
package com.anywherecommerce.android.sdk.sampleapp

import android.app.Activity
import android.graphics.Color
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.anywherecommerce.android.sdk.*
import com.anywherecommerce.android.sdk.AppBackgroundingManager.AppBackroundedListener
import com.anywherecommerce.android.sdk.component.SignatureView
import com.anywherecommerce.android.sdk.devices.CardReader
import com.anywherecommerce.android.sdk.devices.CardReaderController
import com.anywherecommerce.android.sdk.devices.bbpos.BBPOSDevice
import com.anywherecommerce.android.sdk.endpoints.AnyPayTransaction
import com.anywherecommerce.android.sdk.endpoints.worldnet.WorldnetEndpoint
import com.anywherecommerce.android.sdk.models.TipLineItem
import com.anywherecommerce.android.sdk.models.TransactionType
import com.anywherecommerce.android.sdk.sampleapp.MainActivity
import com.anywherecommerce.android.sdk.transactions.CardTransaction
import com.anywherecommerce.android.sdk.transactions.Transaction
import com.anywherecommerce.android.sdk.transactions.listener.CardTransactionListener
import com.anywherecommerce.android.sdk.transactions.listener.TransactionListener
import com.anywherecommerce.android.sdk.util.Amount
import com.bbpos.bbdevice.BBDeviceController

/**
 * Created by Admin on 10/4/2017.
 */
class MainActivity : Activity() {
    var txtPanel: TextView? = null
    var txtReferenceId: EditText? = null
    var btUSBConnect: Button? = null
    var btnConnectAudio: Button? = null
    var btnDisconnectAudio: Button? = null
    var btnIsDeviceConnected: Button? = null
    var btnStartEMV: Button? = null
    var btnConnectBT: Button? = null
    var btnDisconnectBT: Button? = null
    var btnGetTransactions: Button? = null
    var btnTerminalLogin: Button? = null
    var btnKeyedsale: Button? = null
    var btnCaptureTransaction: Button? = null
    var btnRefRefund: Button? = null
    var dialogs = DialogManager()
    var refTransaction: Transaction? = null
    var endpoint: Endpoint? = null
    var propaySessionID: String? = null
    var btnSubmitSignature: Button? = null
    var signatureViewLayout: LinearLayout? = null
    var signatureView: SignatureView? = null
    override fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.testharness_device)
        if (!PermissionsController.verifyAppPermissions(this)) {
            PermissionsController.requestAppPermissions(this, PermissionsController.permissions, 1001)
        }
        Logger.i("SDK Version - " + SDKManager.getSdkVersion())
        Terminal.initialize()
        endpoint = Terminal.getInstance().endpoint
        AppBackgroundingManager.get().registerListener(object : AppBackroundedListener {
            override fun onBecameForeground() {
                Logger.trace("Caught app in foreground.")
            }

            override fun onBecameBackground() {
                Logger.trace("Caught app in background.")
            }
        })
        txtPanel = findViewById<View>(R.id.txtTextHarnessPanel) as TextView
        txtPanel!!.movementMethod = ScrollingMovementMethod()
        btnConnectAudio = findViewById<View>(R.id.audioConnect) as Button
        btnDisconnectAudio = findViewById<View>(R.id.btnDisconnectAudio) as Button
        btnConnectBT = findViewById<View>(R.id.btConnect) as Button
        btnDisconnectBT = findViewById<View>(R.id.btnDisconnectBT) as Button
        btnIsDeviceConnected = findViewById<View>(R.id.btnIsDeviceConnected) as Button
        btnGetTransactions = findViewById<View>(R.id.btnGetTransactions) as Button
        btnStartEMV = findViewById<View>(R.id.btnStartEmvSale) as Button
        btnTerminalLogin = findViewById<View>(R.id.terminalLogin) as Button
        btnKeyedsale = findViewById<View>(R.id.keyedsale) as Button
        btnCaptureTransaction = findViewById<View>(R.id.unrefRefund) as Button
        btnRefRefund = findViewById<View>(R.id.refRefund) as Button
        txtReferenceId = findViewById<View>(R.id.txtReferenceId) as EditText
        btUSBConnect = findViewById<View>(R.id.btUSBConnect) as Button
        signatureViewLayout = findViewById<View>(R.id.signatureViewLayout) as LinearLayout
        signatureView = findViewById<View>(R.id.signatureView) as SignatureView
        btnSubmitSignature = findViewById<View>(R.id.btnSubmitSignature) as Button
        signatureView!!.strokeColor = Color.MAGENTA
        signatureView!!.strokeWidth = 10f
        val cardReaderController = CardReaderController.getControllerFor(BBPOSDevice::class.java)
        cardReaderController.subscribeOnCardReaderConnected { deviceInfo ->
            if (deviceInfo == null) addText("\r\nUnknown device connected") else addText("""

    Device connected ${deviceInfo.modelDisplayName}
    """.trimIndent())
        }
        cardReaderController.subscribeOnCardReaderDisconnected { addText("\nDevice disconnected") }
        cardReaderController.subscribeOnCardReaderConnectFailed { error -> addText("\nDevice connect failed: $error") }
        cardReaderController.subscribeOnCardReaderError { error -> addText("\nDevice error: $error") }
        btUSBConnect!!.setOnClickListener {
            addText("\nConnecting to USB Reader\r\n")
            cardReaderController.connectOther(CardReader.ConnectionMethod.USB)
        }
        btnConnectAudio!!.setOnClickListener {
            addText("\nConnecting to audio jack (with polling)\r\n")
            cardReaderController.connectAudioJack()
        }
        btnDisconnectAudio!!.setOnClickListener {
            addText("\r\nNot implemented")
            //addText("Disconnecting audio jack\r\n");
            //BBPOSDeviceCardReaderController.getInstance().disconnectAudioJack();
        }
        btnConnectBT!!.setOnClickListener {
            addText("\nConnecting to BT\r\n")
            cardReaderController.connectBluetooth { addText("Many BT devices") }
        }
        btnDisconnectBT!!.setOnClickListener {
            addText("Disconnecting bluetooth\r\n")
            cardReaderController.disconnectReader()
        }
        btnIsDeviceConnected!!.setOnClickListener {
            addText("\nClicky")
            BBDeviceController.setDebugLogEnabled(true)
            if (latch) {
                addText("\r\nStopping audio")
                cardReaderController.disconnectReader()
            } else {
                addText("\r\nStarting audio")
                cardReaderController.connectAudioJack()
            }
            latch = !latch
        }
        btnStartEMV!!.setOnClickListener(View.OnClickListener {
            BBDeviceController.setDebugLogEnabled(true)
            addText("\nStarting EMV transaction")
            if (!CardReaderController.isCardReaderConnected()) {
                addText("\r\nNo card reader connected")
                return@OnClickListener
            }
            dialogs.showProgressDialog(this@MainActivity, "Please Wait...")
            sendEMVTransaction()
        })
        btnTerminalLogin!!.setOnClickListener { authenticateTerminal() }
        btnKeyedsale!!.setOnClickListener {
            addText("\r\nExecuting Keyed Sale Transaction. Please Wait...")
            dialogs.showProgressDialog(this@MainActivity, "Please Wait...")
            sendKeyedTransaction()
        }
        btnCaptureTransaction!!.setOnClickListener {
            addText("----> Starting New Refund Transaction <----")
            dialogs.showProgressDialog(this@MainActivity, "Please Wait...")
            val transaction = AnyPayTransaction()
            transaction.endpoint = endpoint
            transaction.transactionType = TransactionType.REFUND
            transaction.totalAmount = Amount("20.25")
            transaction.cardExpiryMonth = "10"
            transaction.cardExpiryYear = "20"
            transaction.address = "123 Main Street"
            transaction.postalCode = "30004"
            transaction.cvV2 = "999"
            transaction.cardholderName = "Jane Doe"
            transaction.cardNumber = "4012888888881881"
            transaction.execute(object : TransactionListener {
                override fun onTransactionCompleted() {
                    dialogs.hideProgressDialog()
                    if (transaction.isApproved) addText("----> Transaction Refunded <----") else {
                        addText(transaction.responseText)
                    }
                }

                override fun onTransactionFailed(meaningfulError: MeaningfulError) {
                    dialogs.hideProgressDialog()
                    addText("Refund Failed")
                }
            })
        }
        btnRefRefund!!.setOnClickListener(View.OnClickListener {
            if (refTransaction == null) {
                addText("----> This voids the last Auth transaction made. Please perform a Auth transaction first. <----")
                return@OnClickListener
            }
            addText("----> Starting Referenced Refund Transaction <----")
            val transaction = AnyPayTransaction()
            transaction.endpoint = endpoint
            transaction.externalId = refTransaction!!.externalId
            transaction.totalAmount = Amount("1")
            transaction.refTransactionId = refTransaction!!.externalId
            transaction.transactionType = TransactionType.REFUND

            //t.enableLogging();
            transaction.execute(object : TransactionListener {
                override fun onTransactionCompleted() {
                    dialogs.hideProgressDialog()
                    if (transaction.isApproved) addText("----> Transaction Refunded <----") else {
                        addText(transaction.responseText)
                    }
                }

                override fun onTransactionFailed(reason: MeaningfulError) {
                    dialogs.hideProgressDialog()
                    addText("Refund Failed")
                }
            })
        })
        btnSubmitSignature!!.setOnClickListener {
            dialogs.showProgressDialog(this@MainActivity, "Please Wait...")
            (refTransaction as CardTransaction?)!!.signature = signatureView!!.signature
            if (refTransaction!!.getCustomField("completed") != null) {
                (endpoint as WorldnetEndpoint?)!!.submitSignature(refTransaction, object : RequestListener<Any?> {
                    override fun onRequestComplete(o: Any?) {
                        addText("\r\n SIgnature Sent Successfully")
                    }

                    override fun onRequestFailed(meaningfulError: MeaningfulError) {
                        addText("\r\n SIgnature sent Failed")
                    }
                })
            } else {
                refTransaction!!.proceed()
            }
            addText("\r\nSending SIgnature")
        }
        authenticateTerminal()
    }

    private fun sendKeyedTransaction() {
        val transaction = AnyPayTransaction()
        transaction.endpoint = endpoint
        transaction.transactionType = TransactionType.SALE
        transaction.cardExpiryMonth = "10"
        transaction.cardExpiryYear = "20"
        transaction.address = "123 Main Street"
        transaction.postalCode = "30004"
        transaction.cvV2 = "999"
        transaction.cardholderName = "Jane Doe"
        transaction.cardNumber = "4012888888881881"
        transaction.totalAmount = Amount("10.47")
        transaction.currency = "USD"
        refTransaction = transaction
        //t.enableLogging();
        transaction.execute(transactionListener)
    }

    private fun sendEMVTransaction() {
        val transaction = AnyPayTransaction()
        transaction.endpoint = endpoint
        transaction.useCardReader(CardReaderController.getConnectedReader())
        transaction.transactionType = TransactionType.SALE
        transaction.address = "123 Main Street"
        transaction.postalCode = "30004"
        transaction.totalAmount = Amount("10.47")
        transaction.currency = "USD"
        transaction.setOnSignatureRequiredListener {
            addText("\r\n------>onSignatureRequired: sending null and proceeding")
            val signature = "base64-encoded image or point map"
            dialogs.hideProgressDialog()
            signatureViewLayout!!.visibility = View.VISIBLE
        }
        refTransaction = transaction

        //t.enableLogging();
        transaction.execute(object : CardTransactionListener {
            override fun onCardReaderEvent(event: MeaningfulMessage) {
                addText("""

    ------>onCardReaderEvent: $event
    """.trimIndent())
            }

            override fun onTransactionCompleted() {
                dialogs.hideProgressDialog()
                addText("""

    ------>onTransactionCompleted${transaction.isApproved}
    """.trimIndent())
                signatureViewLayout!!.visibility = View.GONE
            }

            override fun onTransactionFailed(reason: MeaningfulError) {
                dialogs.hideProgressDialog()
                addText("\r\n------>onTransactionFailed: $reason")
                signatureViewLayout!!.visibility = View.GONE
            }
        })
    }

    private fun addText(text: String) {
        txtPanel!!.append("""

    $text


    """.trimIndent())
    }

    private fun authenticateTerminal() {
        addText("Authenticating. Please Wait...")
        (endpoint as WorldnetEndpoint?)!!.worldnetSecret = ""
        (endpoint as WorldnetEndpoint?)!!.worldnetTerminalID = ""
        (endpoint as WorldnetEndpoint?)!!.url = "https://testpayments.anywherecommerce.com/merchant"
        (endpoint as WorldnetEndpoint?)!!.authenticate(object : AuthenticationListener {
            override fun onAuthenticationComplete() {
                addText("\r\n------> Terminal Authenticated")
            }

            override fun onAuthenticationFailed(meaningfulError: MeaningfulError) {
                addText("\r\n------> Terminal Authentication Failed")
            }
        })
    }

    private val transactionListener: TransactionListener = object : TransactionListener {
        override fun onTransactionCompleted() {
            dialogs.hideProgressDialog()
            addText("""

    ------>onTransactionCompleted - ${refTransaction!!.isApproved}
    """.trimIndent())
            if (refTransaction!!.transactionType == TransactionType.SALE) {
                val tip = TipLineItem()
                tip.amount = Amount("23")
                refTransaction!!.tip = tip
                (endpoint as WorldnetEndpoint?)!!.submitTipAdjustment(refTransaction, tip, object : RequestListener<Any?> {
                    override fun onRequestComplete(o: Any?) {
                        addText("\r\n------>Tip Adjustment success")
                    }

                    override fun onRequestFailed(meaningfulError: MeaningfulError) {
                        addText("\r\n------>Tip Adjustment failed")
                    }
                })

                /*((AnyPayTransaction)refTransaction).update(new TransactionListener() {
                    @Override
                    public void onTransactionCompleted() {
                        addText("\r\n------>update transaction complete - " + refTransaction.isApproved().toString());

                        CustomerDetails customerDetails = new CustomerDetails();
                        customerDetails.setEmailAddress("theiosdevguy@gmail.com");
                        refTransaction.setCustomerDetails(customerDetails);
                        ((AnyPayTransaction) refTransaction).update(new TransactionListener() {
                            @Override
                            public void onTransactionCompleted() {
                                addText("\r\n------>Receipt Sent - ");

                                refTransaction.addCustomField("completed", true);
                                signatureViewLayout.setVisibility(View.VISIBLE);

                            }

                            @Override
                            public void onTransactionFailed(MeaningfulError meaningfulError) {
                                addText("\r\n------>Receipt Sent Failed ");
                            }
                        });
                    }

                    @Override
                    public void onTransactionFailed(MeaningfulError meaningfulError) {
                        addText("\r\n------>update transaction failed - " + refTransaction.isApproved().toString());
                    }
                });*/
            }
        }

        override fun onTransactionFailed(reason: MeaningfulError) {
            dialogs.hideProgressDialog()
            addText("\r\n------>onTransactionFailed: $reason")
        }
    }

    companion object {
        var latch = false
    }
}